### PR TITLE
Add patching for many model config options

### PR DESCRIPTION
**Pull request by tgerdesnv**
Patch a number of functions so that model config info is captured in NVTX ranges:

### CODE REVIEW

File: pyprof/nvtx/nvmarker.py [hunk 1]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+            nvtx.range_pop()
+
+        return result
+
+    setattr(mod, fn_name, wrapper_func)
+
+# Patch functions that help gather high-level configuration options for the model.
+# All resulting nvtx ranges will have 'model_config' as the primary key
+#
+def patch_model_configs():
```
Comment (tgerdesnv) [inline pyprof/nvtx/nvmarker.py:461]: Any idea how to test/protect any of this? I've manually tested most of it.

**Comment by adityaiitb** [comment]
hi,

Thanks for using the tool and the pull request. I would definitely like to understand this better. Can you explain in a little more detail:
- What is the problem this request is trying to solve?
- What is the typical solution i.e. how do users investigate the issue?
- What information is being captured to solve the problem?
- Does this help with performance analysis?

One can use monkey patching to do a lot of tricks e.g. add NVTX annotations, read / modify tensor values, inject code etc. I looked at the code. I didn't understand the intent but seems useful. However, I am not sure if its related to performance analysis. It appears, your solution (for the problem you encountered) could be a separate library in its own right. Please help us understand the problem you faced/solved better.

File: pyprof/nvtx/nvmarker.py [hunk 2]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+# Flag to indicate that cudnn_benchmark_disabled has already been reported
+#
+global cudnn_benchmark_disabled_reported
+cudnn_benchmark_disabled_reported = False
+
+# Path the given mod/function so that if it is ever executed and 
+# torch.backends.cudnn.benchmark is not true, it will emit an nvtx
+# range to report that fact
+#
+def patch_with_always_benchmark(mod, fn_name):
```
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:390]: According to this: https://discuss.pytorch.org/t/what-does-torch-backends-cudnn-benchmark-do/5936
we only want to enable this when the input sizes are not varying. Are we accounting for that anywhere before we set this?

File: pyprof/nvtx/nvmarker.py [hunk 3]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+            nvtx.range_pop()
+
+        return result
+
+    setattr(mod, fn_name, always_benchmark_wrapper)
+
+
+# Patch the given mod/function. If the function is executed, emit 
+# an nvtx_range with data indicating that 'key' was true
+#
```
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:414]: Should be docstrings.

File: pyprof/nvtx/nvmarker.py [hunk 4]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+        nvtx.range_pop()
+        return result
+
+    setattr(mod, fn_name, wrapper_func)
+
+
+# Patch the given mod/function. If the function is executed 
+# and any of the bad args have any of the listed bad values, 
+# emit an nvtx_range with data indicating that 'key' was true
+#
```
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:430]: Should be docstrings.

File: pyprof/nvtx/nvmarker.py [hunk 5]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+
+    def wrapper_func(*args, **kwargs):
+
+        signature = ins.signature(old_fn)
+        bound = signature.bind(*args, **kwargs)
+        bound.apply_defaults()
+
+        problem = False
+        for k,v in bound.arguments.items():
+            if (k in bad_args):
```
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:442]: do not need parentheses.

File: pyprof/nvtx/nvmarker.py [hunk 6]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+                if (v in bad_args[k]):
```
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:443]: no paren

File: pyprof/nvtx/nvmarker.py [hunk 7]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+                    problem = True
+
+        if (problem):
```
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:446]: no paren

File: pyprof/nvtx/nvmarker.py [hunk 8]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+            push_nvtx_model_config({key: True})
+
+        result = old_fn(*args, **kwargs)
+
+        if (problem):
```
Comment (debermudez) [inline pyprof/nvtx/nvmarker.py:451]: no paren

File: pyprof/nvtx/nvmarker.py [hunk 9]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
         import fused_layer_norm_cuda
         patchClass(fused_layer_norm_cuda)
 
+
+# Helper function to dump the passed in dict config as an nvtx
+# marker with "model_config" key
+#
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py [hunk 10]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+def push_nvtx_model_config(config):
+    nvtx_msg = json.dumps({"model_config": config})
+    nvtx.range_push(nvtx_msg)
+
+
+# Capture dataloader config (num_workers and pin_memory) and
+# emit a model_config nvtx range with the information
+#
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py [hunk 11]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+
+        push_nvtx_model_config({"num_workers": num_workers, "pin_memory": pin_memory})
+        old_init(self, *args, **kwargs)
+        nvtx.range_pop()
+
+    mod.DataLoader.__init__ = new_init
+
+
+# Flag to indicate that cudnn_benchmark_disabled has already been reported
+#
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py:380]: done

File: pyprof/nvtx/nvmarker.py [hunk 12]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+global cudnn_benchmark_disabled_reported
+cudnn_benchmark_disabled_reported = False
+
+# Path the given mod/function so that if it is ever executed and 
+# torch.backends.cudnn.benchmark is not true, it will emit an nvtx
+# range to report that fact
+#
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py [hunk 13]
```diff
@@ -350,12 +350,141 @@ def patch_apex():
...
+        push_nvtx_model_config({"num_workers": num_workers, "pin_memory": pin_memory})
+        old_init(self, *args, **kwargs)
+        nvtx.range_pop()
+
+    mod.DataLoader.__init__ = new_init
+
+
+# Flag to indicate that cudnn_benchmark_disabled has already been reported
+#
+global cudnn_benchmark_disabled_reported
```
[the remark this answers is not in the document]
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py (hunk 3 above)
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py (hunk 4 above)
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py (hunk 5 above)
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py (hunk 6 above)
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py (hunk 7 above)
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py (hunk 8 above)
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py]: done

File: pyprof/nvtx/nvmarker.py (hunk 2 above)
Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py:383]: We aren't accounting for that here. We can handle it in the script that consumes this info.

Comment (debermudez) [reply pyprof/nvtx/nvmarker.py:383]: I understand now. Got ahead of myself

Comment (tgerdesnv) [reply pyprof/nvtx/nvmarker.py:383]: No, it was an excellent point that I wasn't aware of. Regardless, no change needed here.

File: pyprof/nvtx/nvmarker.py (hunk 1 above)
Comment (debermudez) [reply pyprof/nvtx/nvmarker.py:459]: Not really.

**Pull request merged**